tests/kafkatest/services/console_consumer.py

```diff
@@ -226,10 +230,13 @@ def _worker(self, idx, node):
 
             for line in itertools.chain([first_line], consumer_output):
                 msg = line.strip()
-                if self.message_validator is not None:
-                    msg = self.message_validator(msg)
-                if msg is not None:
-                    self.messages_consumed[idx].append(msg)
+                if msg == "shutdown_complete":
+                    self.nodes_clean_shutdown.append(idx)
+                else:
+                    if self.message_validator is not None:
+                        msg = self.message_validator(msg)
+                    if msg is not None:
+                        self.messages_consumed[idx].append(msg)
```

Comment:
Agree about a set. Based on java consumer code, should be called just once. I guess it's always better to catch weird behavior earlier -- I will raise an exception on a duplicate shutdown event.